### ISSUE TITLE
track_order not working on root group

### ISSUE THREAD

**Issue by zonca** [h5py/h5py#1577]
This is a duplicate of https://github.com/h5py/h5py/issues/1097,
it was closed but I still experience this issue.

```
import tempfile
import h5py
print(h5py.version.info)
ft = tempfile.mktemp(suffix='.hdf5')
h5py.get_config().track_order = True
h5 = h5py.File(ft, mode='w', track_order=True)
h5.create_group('z')
h5.create_group('a')
print("File just created", h5.keys())
h5.close()
h5 = h5py.File(ft, mode='r', track_order=True)
print("After reopening", h5.keys())
```

Output:
```
Summary of the h5py configuration
---------------------------------

h5py    2.10.0
HDF5    1.10.5
Python  3.7.0 | packaged by conda-forge | (default, Nov 12 2018, 20:15:55) 
[GCC 7.3.0]
sys.platform    linux
sys.maxsize     9223372036854775807
numpy   1.17.5

File just created <KeysViewHDF5 ['z', 'a']>
After reopening <KeysViewHDF5 ['a', 'z']>
```

**Comment by takluyver** [comment]
cc @aparamon

**Comment by dota17** [comment]
I would say that this is a bug.

I found that:
1. all group(root group or subgroup) are stored in alphabetical order
2. the setting `track_order`  wheather in read mode or write mode has no effect if we set it globally via h5py.get_config()

**Comment by takluyver** [comment]
To be clear, it's not expected to do anything in read mode - order tracking has to be on when you create the objects, otherwise HDF5 doesn't store the order information.

**Comment by dota17** [comment]
Yeah, I understand. User can just remove that part when reading the file, like h5py.File(ft, mode='r'~~, track_order=True~~)

**Comment by takluyver** [comment]
Possibly a duplicate of #1471.

**Comment by dota17** [comment]
I think I have figured it out why this happened.
When we are in `r` mode:

in `make_fid()` :
https://github.com/h5py/h5py/blob/4016dab95dc853d3f18e1c7df22a10cb5d229ed7/h5py/_hl/files.py#L186-L196




the `fcpl` produced by `make_fcpl`(line 425) is useless, because `h5f.open()`(line 190) doesn't accept `fcpl`, but `h5f.create()`(line 196) do, so in `w` mode, the `track_order` in `make_fcpl()` works.

https://github.com/h5py/h5py/blob/4016dab95dc853d3f18e1c7df22a10cb5d229ed7/h5py/_hl/files.py#L422-L427


so even though we set it like this:
```python
def make_fcpl(track_order=False, fs_strategy=None, fs_persist=False, fs_threshold=1):
    """ Set up a file creation property list """
    if track_order or fs_strategy:
        plist = h5p.create(h5p.FILE_CREATE)
        if track_order:
            plist.set_link_creation_order(
                h5p.CRT_ORDER_TRACKED | h5p.CRT_ORDER_INDEXED)
            plist.set_attr_creation_order(
                h5p.CRT_ORDER_TRACKED | h5p.CRT_ORDER_INDEXED)
```
when we read again, nothing will change for root group

**Comment by takluyver** [comment]
I don't quite follow what you're saying. Yes, `track_order` doesn't do anything when you're opening an existing file. It needs to be set when the file is created. The bug is that it apparently doesn't work even if you set it when creating a file.

**Comment by dota17** [comment]
But as I can see, before closing the file, the track order works as we expected:
```python
h5 = h5py.File(ft, mode='w', track_order=True)
h5.create_group('z')
h5.create_group('a')
print("File just created", h5.keys())
```
output is `<KeysViewHDF5 ['z', 'a']>`

**Comment by takluyver** [comment]
I believe the idea is that the order is saved in the file, so if you set `track_order=True` when creating the file, the order is still used if you reopen the same file.

**Comment by vasole** [comment]
Please, try to write h5["/"].keys() instead of h5.keys(). The former works as expected for me but not the latter.

**Comment by mgarort** [comment]
@vasole I come a bit late to this conversation but I confirm that for me `h5["/"].keys()` keeps the order but `h5.keys()` does not.

**Comment by takluyver** [comment]
It's been pointed out that there's a simple workaround - replace `f.keys()` with `f["/"].keys()` and it does remember the order. But yes, it is still a bug. I think it's because this line:

https://github.com/h5py/h5py/blob/71af64dec871ed3545b18c6d4d17d9b0c00032e5/h5py/h5g.pyx#L120

gets a file creation property list (instead of group creation). That must not include the group creation information for the root group.

**Comment by ajelenak** [comment]
Perhaps a fix could be to have an `__iter__` method for the `File` class like the one for the `Group` class, and in there use `self['/']`?

**Comment by takluyver** [comment]
Thanks @ajelenak - I've now done basically that, but one level lower, in #2410.
